pandas/core/arrays/text.py

```diff
@@ -0,0 +1,277 @@
+        return cls._from_sequence(strings, dtype=dtype, copy=copy)
+
+    def __setitem__(self, key, value):
+        value = extract_array(value, extract_numpy=True)
+        if isinstance(value, type(self)):
+            # extract_array doesn't extract PandasArray subclasses
+            value = value._ndarray
+
+        scalar_key = lib.is_scalar(key)
+        scalar_value = lib.is_scalar(value)
+        if scalar_key and not scalar_value:
+            raise ValueError("setting an array element with a sequence.")
+
+        # validate new items
+        if scalar_value:
+            if scalar_value is None:
+                value = np.nan
+            elif not (isinstance(value, str) or np.isnan(value)):
+                raise ValueError(
+                    "Cannot set non-string value '{}' into a TextArray.".format(value)
+                )
+        else:
+            if not is_array_like(value):
+                value = np.asarray(value, dtype=object)
+            if len(value) and not lib.is_string_array(value, skipna=True):
+                raise ValueError("Must provide strings.")
+
+        super().__setitem__(key, value)
+
+    def fillna(self, value=None, method=None, limit=None):
+        # TODO: validate dtype
+        return super().fillna(value, method, limit)
+
+    def astype(self, dtype, copy=True):
+        dtype = pandas_dtype(dtype)
+        if isinstance(dtype, TextDtype):
+            if copy:
+                return self.copy()
+            return self
+        return super().astype(dtype, copy)
+
+    def _reduce(self, name, skipna=True, **kwargs):
+        raise TypeError("Cannot perform reduction '{}' with text dtype".format(name))
+
+    def value_counts(self, dropna=False):
+        from pandas import value_counts
+
+        return value_counts(self._ndarray, dropna=dropna)
+
+    # Overrride parent because we have different return types.
+    @classmethod
+    def _create_arithmetic_method(cls, op):
+        def method(self, other):
+            if isinstance(other, (ABCIndexClass, ABCSeries, ABCDataFrame)):
+                return NotImplemented
+
+            elif isinstance(other, cls):
+                other = other._ndarray
+
+            mask = isna(self) | isna(other)
+            valid = ~mask
+
+            if not lib.is_scalar(other):
+                if len(other) != len(self):
+                    # prevent improper broadcasting when other is 2D
+                    raise ValueError(
+                        "Lengths of operands do not match: {} != {}".format(
+                            len(self), len(other)
+                        )
+                    )
+
+                other = np.asarray(other)
+                other = other[valid]
+
+            result = np.empty_like(self._ndarray, dtype="object")
+            result[mask] = np.nan
+            result[valid] = op(self._ndarray[valid], other)
+
+            if op.__name__ in {"add", "radd", "mul", "rmul"}:
```

Comment:
are there tests for these?

Reply:
Yes. `test_add` and `test_mul` in the new `test_text.py` file.